src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -642,5 +688,100 @@ public static ClusterState readFrom(StreamInput in, @Nullable DiscoveryNode loca
             }
             return builder.build();
         }
+
+    }
+
+    @Override
+    public Diff diff(ClusterState previousState) {
+        return new ClusterStateDiff(previousState, this);
+    }
+
+    @Override
+    public ClusterState readFrom(StreamInput in) throws IOException {
+        return Builder.readFrom(in, null);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        Builder.writeTo(this, out);
+    }
+
+    private static class ClusterStateDiff implements Diff {
+
+        private final long version;
+
+        private final String previousUuid;
+
+        private final String uuid;
+
+        private final ClusterName clusterName;
+
+        private final Diff routingTable;
+
+        private final Diff nodes;
+
+        private final Diff metaData;
+
+        private final Diff blocks;
+
+        private final Diff customs;
+
+        public ClusterStateDiff(ClusterState before, ClusterState after) {
```

Comment:
Can we call these `from` and `to` and the non Diff fields `toVersion` `fromUuid` and `toUuid` ? sorry to be a  pain but I think it will be clearer.